util/parallel.cpp

```diff
@@ -0,0 +1,174 @@
+      ENSURE(close(c.pipe[0]) == 0);
+  } else {
+    /*
+     * parent -- close the write side of the new pipe and mark the
+     * read side as non-blocking
+     */
+    ENSURE(close(newKid.pipe[1]) == 0);
+    int flags = fcntl(newKid.pipe[0], F_GETFL, 0);
+    assert(flags != -1);
+    ENSURE(fcntl(newKid.pipe[0], F_SETFL, flags | O_NONBLOCK) != -1);
+    ++subprocesses;
+    newKid.pid = pid;
+  }
+  return { pid, &newKid.output, index };
+}
+
+/*
+ * return true only if all children have returned EOF
+ */
+bool parallel::readFromChildren() {
+  const int maxRead = 4096;
+  static char data[maxRead];
+  bool allEOF = true;
+  for (auto &c : children) {
+    if (c.eof)
+      continue;
+    allEOF = false;
+  again:
+    size_t size = read(c.pipe[0], data, maxRead);
+    if (size == (size_t)-1) {
+      assert(errno == EAGAIN);
+      continue;
+    }
+    if (size == 0) {
+      c.eof = true;
+      continue;
+    }
+    auto *pbuf = c.output.rdbuf();
+    ENSURE((size_t)pbuf->sputn(data, size) == size);
+    /*
+     * keep reading from this pipe until there's nothing left -- we
+     * want to minimize the time TV processes spend blocked
+     */
+    goto again;
+  }
+  return allEOF;
+}
+
+/*
+ * wrapper for write() that correctly handles short writes
+ */
+static ssize_t safe_write(int fd, const void *void_buf, size_t count) {
+  const char *buf = (const char *)void_buf;
+  ssize_t written = 0;
+  while (count > 0) {
+    ssize_t ret = write(fd, buf, count);
+    // let caller deal with EOF and error conditions
+    if (ret <= 0)
+      return ret;
+    count -= ret;
+    buf += ret;
+    written += ret;
+  }
+  return written;
+}
+
+void parallel::writeToParent() {
+  ensureChild();
+  childProcess &me = children.back();
+  std::streambuf *pbuf = me.output.rdbuf();
+  std::streamsize size = pbuf->pubseekoff(0, me.output.end);
+  pbuf->pubseekoff(0, me.output.beg);
+  char *data = new char[size];
+  pbuf->sgetn(data, size);
+  ENSURE(safe_write(me.pipe[1], data, size) == size);
+  delete[] data;
+}
+
+void parallel::ensureParent() {
+  assert(parent_pid != -1 && getpid() == parent_pid);
+}
+
+void parallel::ensureChild() {
+  assert(parent_pid != -1 && getpid() != parent_pid);
+}
+
+bool parallel::init(int _max_subprocesses) {
+  assert(parent_pid == -1);
+  parent_pid = getpid();
+  max_subprocesses = _max_subprocesses;
+  return true;
+}
+
+void parallel::waitForAllChildren() {
+  ensureParent();
+  // FIXME: we could use poll() instead of this polling loop
+  const struct timespec delay = {0, 100 * 1000 * 1000}; // 100ms
+  while (!readFromChildren())
+    nanosleep(&delay, 0);
+  int status;
+  while (wait(&status) != -1) {
+    if (WIFEXITED(status))
+      --subprocesses;
+  }
+  assert(subprocesses == 0);
+}
+
+void parallel::emitOutput(std::stringstream &parent_ss,
+                          std::ofstream &out_file) {
+  ensureParent();
+  std::string line;
+  std::regex rgx("^include\\(([0-9]+)\\)$");
```

Comment:
is this bazooka really needed? Isn't a string_view::starts_with() sufficient?

Reply:
I can revise if you want but I like the way it expresses all of the constraints at once instead of spreading them across a few lines of less-obvious code. 

also we're only going to do this operation once per fork + TV call, so it's hardly plausible that this would show up in our profile.